File: snippets/python/parse-pagination-query-params-safely-with-python/src/parse_pagination_query_params.py
from dataclasses import dataclass
from typing import Mapping


DEFAULT_PAGE = 1
DEFAULT_PAGE_SIZE = 20
MAX_PAGE_SIZE = 100


@dataclass(frozen=True)
class PaginationParams:
    page: int
    page_size: int
    offset: int


def parse_pagination_query_params(query_params: Mapping[str, str]) -> PaginationParams:
    """Parse page and page_size query params with defaults and limits."""
    # Flow:
    #   read page and page_size from the query mapping
    #      |
    #      +-> valid integers -> apply defaults and max limit -> return params
    #      `-> invalid or out-of-range values -> raise ValueError
    page = _read_positive_int(query_params, "page", default=DEFAULT_PAGE)
    page_size = _read_positive_int(query_params, "page_size", default=DEFAULT_PAGE_SIZE)
    bounded_page_size = min(page_size, MAX_PAGE_SIZE)
    offset = (page - 1) * bounded_page_size
    return PaginationParams(page=page, page_size=bounded_page_size, offset=offset)


def _read_positive_int(query_params: Mapping[str, str], key: str, default: int) -> int:
    raw_value = query_params.get(key)
    if raw_value is None:
        return default
    if not isinstance(raw_value, str):
        raise ValueError(f"{key} must be a string integer")

    try:
        value = int(raw_value.strip())
    except ValueError as error:
        raise ValueError(f"{key} must be a valid integer") from error

    if value < 1:
        raise ValueError(f"{key} must be greater than or equal to 1")
    return value


if __name__ == "__main__":
    params = parse_pagination_query_params({"page": "3", "page_size": "50"})
    print(params)

File: snippets/go/parse-pagination-query-params-safely-with-go/parse_pagination_query_params.go
package main

import (
	"errors"
	"fmt"
	"log"
	"net/url"
	"strconv"
	"strings"
)

type Pagination struct {
	Page     int
	PageSize int
	Offset   int
}

// Flow: read page and page_size -> apply defaults and validation -> cap the page size -> return page, size, and offset.
func ParsePaginationQueryParams(values url.Values, defaultPage, defaultPageSize, maxPageSize int) (Pagination, error) {
	if defaultPage <= 0 {
		return Pagination{}, errors.New("default page must be greater than 0")
	}
	if defaultPageSize <= 0 {
		return Pagination{}, errors.New("default page size must be greater than 0")
	}
	if maxPageSize <= 0 {
		return Pagination{}, errors.New("max page size must be greater than 0")
	}

	page, err := parsePositiveIntOrDefault(values.Get("page"), defaultPage)
	if err != nil {
		return Pagination{}, fmt.Errorf("parse page: %w", err)
	}

	pageSize, err := parsePositiveIntOrDefault(values.Get("page_size"), defaultPageSize)
	if err != nil {
		return Pagination{}, fmt.Errorf("parse page size: %w", err)
	}
	if pageSize > maxPageSize {
		pageSize = maxPageSize
	}

	return Pagination{
		Page:     page,
		PageSize: pageSize,
		Offset:   (page - 1) * pageSize,
	}, nil
}

func parsePositiveIntOrDefault(raw string, fallback int) (int, error) {
	trimmed := strings.TrimSpace(raw)
	if trimmed == "" {
		return fallback, nil
	}

	value, err := strconv.Atoi(trimmed)
	if err != nil {
		return 0, errors.New("must be an integer")
	}
	if value <= 0 {
		return 0, errors.New("must be greater than 0")
	}

	return value, nil
}

func main() {
	values := url.Values{
		"page":      {"2"},
		"page_size": {"100"},
	}

	pagination, err := ParsePaginationQueryParams(values, 1, 25, 50)
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("page=%d page_size=%d offset=%d", pagination.Page, pagination.PageSize, pagination.Offset)
}
